### PR TITLE
feature/template

### PR DESCRIPTION
**Pull request by sprzesmycki**

### CODE REVIEW

File: awesome_lottery/main.py [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    winners = []
+    for i in range(rewards.get_prizes_count()):
+        winner = get_winner(participants)
+        winners.append(winner)
+        reduce_winner_chances(winner)
+
+    lottery_data = Lottery(participants=participants, prizes=prizes, winners=winners)
+    lottery_json = Lottery.to_json(lottery_data)
+    print(lottery_json)
+    winners_with_prizes = get_winners_with_prizes(winners, rewards)
```
Comment (GrivIN) [inline awesome_lottery/main.py:179]: whole this part should be a part of lottery class

File: awesome_lottery/main.py [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+    if results is None:
+        winners_with_prizes.print_winners()
+    else:
+        with open(results, 'w') as f:
+            f.write(Winners.to_json(winners_with_prizes))
```
Comment (GrivIN) [inline awesome_lottery/main.py:185]: this logic should be split into 2 classes StdoutWriter (prints to screen) and FileWriter (writes to file as json)

File: awesome_lottery/main.py [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+    return Winners(winners=list_of_winners_with_prizes)
+
+
+def get_first_file_in_path(path):
+    return next(Path(f"{path}/").iterdir())
+
+
+@click.command()
+@click.option('--file', help='Provide file path to process')
+@click.option('--filetype', help='Provide file type csv/json')
```
Comment (GrivIN) [inline awesome_lottery/main.py:152]: https://click.palletsprojects.com/en/8.0.x/options/#choice-options is the best type here

File: awesome_lottery/main.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+@click.option('--results', help='Provide filename to save results')
+def lottery(file, filetype, rewards, results):
+    participants = []
+    if file is not None:
+        filetype = filetype or get_extension(file)
+        participants = read_participants_file(file, filetype)
+    else:
+        print('-- File not specified --')
+        exit()
+    participants = get_list_of_participants(participants)
```
Comment (GrivIN) [inline awesome_lottery/main.py:163]: you can use different approach: https://click.palletsprojects.com/en/8.0.x/options/#callbacks-for-validation

File: awesome_lottery/main.py [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                i.amount -= 1
+                return i
+
+
+@dataclass_json
+@dataclass
+class Winner:
+    first_name: str
+    last_name: str
+    prize: str
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:52]: Fixed

File: awesome_lottery/main.py [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+
+
+@dataclass_json
+@dataclass
+class Winners:
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:57]: I'm using this to generate output.json
`f.write(Winners.to_json(winners_with_prizes))`
Should I use different way to save this to file?

File: awesome_lottery/main.py [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+    id: int
+    name: str
+    amount: 1
+
+
+@dataclass_json
+@dataclass
+class Prizes:
+    name: str
+    prizes: list[Prize]
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:32]: What difference is between List and list?
Screenshot 2021-12-30 at 22 08 03

File: awesome_lottery/main.py [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+    def get_prizes_count(self):
+        for i in self.prizes:
+            self.total_count += i.amount
+        return self.total_count
+
+    def get_next_prize(self):
+        for i in self.prizes:
+            if i.amount > 0:
+                i.amount -= 1
+                return i
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:44]: Should it be used that way?
```
def get_winners_with_prizes(winners, prizes: Prizes):
    list_of_winners_with_prizes = []
    iter__ = prizes.__iter__()
    for winner in winners:
        list_of_winners_with_prizes.append(
            Winner(participant=winner, prize=next(iter__)))
    return Winners(winners=list_of_winners_with_prizes)
```

File: awesome_lottery/main.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+@dataclass_json
+@dataclass
+class Lottery:
+    participants: list[Participant]
+    prizes: list[Prizes]
+    winners: list[Winner]
+
+
+def read_file(file_path, file_type):
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:75]: read_participants_file iterates over participants and manipulates data
read_file returns data as it is

File: awesome_lottery/main.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+    for row in reader:
+        weight = row.get('weight')
+        if weight is None:
+            weight = 1
+        else:
+            weight = int(weight)
+        yield Participant(int(row.get('id')),
+                          row.get('first_name'),
+                          row.get('last_name'),
+                          weight)
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:111]: fixed

File: awesome_lottery/main.py [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+
+def get_winner(list_of_users):
+    list_of_weights = list(map(lambda user: user.weight, list_of_users))
+    return random.choices(list_of_users, weights=list_of_weights, k=1)[0]
+
+
+def reduce_winner_chances(winner):
+    winner.weight -= 1
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:120]: fixed

File: awesome_lottery/main.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (GrivIN) [reply awesome_lottery/main.py:32]: o nawet nie wiedziałem 👍🏻

File: awesome_lottery/main.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (GrivIN) [reply awesome_lottery/main.py:44]: no, you should never call `__xxx__()` functions directly

File: awesome_lottery/main.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (GrivIN) [reply awesome_lottery/main.py:57]: there is `json.dump[s]` functions, but i do not have strong opinion about the best way to do it, leaving in your jurisdiction ;)

File: awesome_lottery/main.py [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+
+def get_list_of_participants(participants):
+    list_of_participants = []
+    it = iter(participants)
+    while True:
+        try:
+            list_of_participants.append(next(it))
+        except StopIteration:
+            break
+    return list_of_participants
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:135]: Fixed - didn't knew its possible to do list of yield

File: awesome_lottery/main.py [hunk 13]
```diff
@@ -0,0 +1,189 @@
...
+@dataclass
+class Prizes:
+    name: str
+    prizes: list[Prize]
+    total_count: int = 0
+
+    def get_prizes_count(self):
+        for i in self.prizes:
+            self.total_count += i.amount
+        return self.total_count
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:38]: fixed

File: awesome_lottery/main.py [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+        else:
+            weight = int(weight)
+        yield Participant(int(row.get('id')),
+                          row.get('first_name'),
+                          row.get('last_name'),
+                          weight)
+
+
+def get_winner(list_of_users):
+    list_of_weights = list(map(lambda user: user.weight, list_of_users))
```
[the remark this answers is not in the document]
Comment (sprzesmycki) [reply awesome_lottery/main.py:115]: fixed

File: awesome_lottery/main.py (hunk 3 above)
Comment (sprzesmycki) [reply awesome_lottery/main.py:152]: fixed

File: awesome_lottery/main.py (hunk 4 above)
Comment (sprzesmycki) [reply awesome_lottery/main.py]: fixed but I assume it should be simplified.

File: awesome_lottery/main.py [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+    def __iter__(self):
+        for participant in self.winners:
+            for _ in range(participant.amount):
+                yield participant
+
+    def __str__(self):  # TODO not sure if this shouldn't be moved to lottery class??
+        print('Winners of lottery: ')
+        for winner in self.winners:
+            print(f'-> {winner.participant.first_name} {winner.participant.last_name} won {winner.prize.name}')
+        print('Congratulations for winners!')
```
Comment (GrivIN) [inline awesome_lottery/main.py:70]: this function should not use `print()` method. Its intend is to return a string that can be printed

File: awesome_lottery/main.py [hunk 16]
```diff
@@ -0,0 +1,174 @@
...
+class Lottery:
+    participants: List[Participant]
+    prizes: List[Prizes]
+    winners: List[Participant]
+
+    def get_winners(self):
+        for i in range(self.prizes.total_count):
+            winner = self.get_winner()
+            self.winners.append(winner)
+            winner.reduce_win_change()
```
Comment (GrivIN) [inline awesome_lottery/main.py:84]: those 2 should be a part of get_winner() method, Lets discuss why

**Pull request merged**